### PR TITLE
Optimised clipping in clahe and some other minor optimisations

### PR DESCRIPTION
**Pull request by lferraz**
### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/enhance/equalization.py [hunk 1]
```diff
@@ -112,36 +112,26 @@ def _compute_luts(tiles_x_im: torch.Tensor, num_bins: int = 256, clip: float = 4
...
-                hist = torch.where(idxs, clip_limit, hist)
-
-                redist: float = clipped // num_bins
-                hist += redist
-
-                residual: float = clipped - redist * num_bins
-                if residual:
-                    hist[0:int(residual)] += 1
-                histos[i] = hist
+        histos.clamp_(max=max(clip * pixels // num_bins, 1))
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:126]: In place operations won’t work well with auto grad

File: kornia/enhance/equalization.py [hunk 2]
```diff
@@ -112,36 +112,26 @@ def _compute_luts(tiles_x_im: torch.Tensor, num_bins: int = 256, clip: float = 4
...
+        clipped: torch.Tensor = pixels - histos.sum(1)
+        redist: torch.Tensor = clipped // num_bins
+        histos += redist[None].T
+        residual: torch.Tensor = clipped - redist * num_bins
+        # trick to avoid using a loop to assign the residual
+        v_range: torch.Tensor = torch.arange(num_bins, device=histos.device)
+        mat_range: torch.Tensor = v_range.repeat(histos.shape[0], 1)
+        histos += mat_range < residual[None].T
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:134]: .T won’t work with old pytorch versions

File: kornia/enhance/equalization.py [hunk 3]
```diff
@@ -241,27 +230,27 @@ def _compute_equalized_tiles(interp_tiles: torch.Tensor, luts: torch.Tensor) ->
...
+    b = torch.addcmul(br, tiw, torch.sub(bl, br))
+    tiles_equalized[:, 1:-1, 1:-1] = torch.addcmul(b, tih, torch.sub(t, b))
 
     # corner regions
     tiles_equalized[:, 0::gh - 1, 0::gw - 1] = preinterp_tiles_equalized[:, 0::gh - 1, 0::gw - 1, 0]
 
     # border region (h)
     t, b, _, _ = preinterp_tiles_equalized[:, 1:-1, 0].unbind(2)
-    tiles_equalized[:, 1:-1, 0] = tih.squeeze(1) * (t - b) + b
+    tiles_equalized[:, 1:-1, 0] = torch.addcmul(b, tih.squeeze(1), torch.sub(t, b))
```
Comment (edgarriba) [inline kornia/enhance/equalization.py:242]: Is addcmul faster ?

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/enhance/equalization.py (hunk 3 above)
Comment (lferraz) [reply kornia/enhance/equalization.py:242]: Yes, because the same aten function is running both ops. At least with nsight I saw a noticeable change.

Comment (lferraz) [reply kornia/enhance/equalization.py:242]: I tried to use JIT to see if all the ops could be joined and computed together in the same cuda thread but it was a waste of time :(. If we have an expert in optimisation at Kornia we could discuss a little bit about these things. :)

**Review by edgarriba — APPROVED** [review]

**Review by ducha-aiki — APPROVED** [review]

File: kornia/enhance/equalization.py (hunk 3 above)
Comment (edgarriba) [reply kornia/enhance/equalization.py:245]: @t-vi any insights?
